compiler/main/checks.cpp

```diff
@@ -180,6 +180,58 @@ void check_flattenFunctions()
   // Suggestion: Ensure no nested functions.
 }
 
+// copied and modified from `symExprIsUsedAsConstRef` in `cullOverReferences`
```

Comment:
I appreciate that you added the comment (it's certainly preferable to not having it). Though I have to say (and apologies for the snarkyness) but I can't help but read it as: "I'm introducing code duplication and I'm introducing it across files."

The day `symExprInUsedAsRef` gets modified (maybe to add a new case, or optimize it, or fix a bug) there's a risk the same won't be applied to `symExprIsUsedAsConstRef`. Adding a cross-referencing comment at `symExprIsUsedInRef` makes that a little bit less likely to happen; though it's all too easy to gloss over comments. So it'd be be better if we code avoid code duplication across files and even better(er?) if we could avoid code duplication all-together.

Taking a look at it, if I strip out comments look at the diff of the two versions it doesn't seem like it'd be that impossible to unify them:

![image](https://github.com/chapel-lang/chapel/assets/2591890/c59d194c-fdd5-463f-ba4a-37b4484f252f)

I could imagine having both call out to a more general `symExprIsUsedAsRefImpl` function that either:

- merges the two and add an input parameter to switch to the desired behavior at runtime: add a `bool processConstRef` parameter and uses if/else as needed where the two existing implementations diverge, or
- use dependency injection:
  - the functional way: pass in `processSymExprUnderFunctionCall`, `processSymexprUnderReturnOrYield` functions and then call out as appropriate.
  - or the OO way: wrap the function into an abstract class, add virtual functions for callbacks to the differing behavior, create two concrete subclasses that implement the different behavior.

Though if we do try and unify them it raises the question about where to put the unified version.

As a more general gripe: it seems like we have a lot of these utility/tool functions spread throughout various passes in the production compiler (I've been guilty of writing some myself) and it'd be nice if we could try and move the more general utility-level things into more general utility-level files.

One problem I have is I'm really not sure who to consult when it comes to organization of the code in the production compiler. @bradcray, any thoughts on that? Pinging Brad every time I have a question about "where should I put X in the production compiler" doesn't seem scalable but if it's not Brad, I'm not sure who is considered our "expert/leader" on the architecture of the production compiler.

Though I suppose, there's also a question of how general this utility is. If it's really embedding a lot of assumptions about the state of the AST after `cullOverReferences` then maybe it'd be best to have declaration to
`symExprIsUsedAsRef` and `symExprIsUsedAsConstRef` in something like `cullOverReferences.h`. If not then maybe we'd want it in some more general  `SymUtils.h`/`SymTools.h` file or maybe on `symbol.h` itself?

A bit of a tangent on the original code (feel free to ignore this if you don't happen to know the answer):

In the original code what's up with us checking  `if (lhsSymbol->hasFlag(FLAG_REF_VAR))` and then asserting or returning and then later we check `lhsSymbol->isRef()`. Does `lhsSymbol->isRef()` not imply `lhsSymbol->hasFlag(FLAG_REF_VAR)`?

In your version you took out `lhsSymbol->isRef()`, why? Is that safe?

Reply:
I will see what I can do about the code duplication, pending further discussion on where to put a combined function.

> In the original code what's up with us checking if (lhsSymbol->hasFlag(FLAG_REF_VAR)) and then asserting or returning and then later we check lhsSymbol->isRef(). Does lhsSymbol->isRef() not imply lhsSymbol->hasFlag(FLAG_REF_VAR)?

I can only hazard a guess that its for some historical reason...

> In your version you took out lhsSymbol->isRef(), why? Is that safe?

This is so `ref a = A[B];` does NOT warn. This is a lifetime error and it seemed silly to me to have a deprecation warning on an error.

Reply:
I extracted out the functionality into `compiler/include/astutil.h`, see https://github.com/chapel-lang/chapel/pull/23208/commits/d82a2ff6162077f590f0c1ffee6ef657df23fc97. Running a paratest with/out gasnet now to make sure everything still works

Reply:
I love it, thanks for taking the time to refactor this.